.pre-commit-config.yaml

```diff
@@ -205,6 +207,35 @@ repos:
     - --namespace-packages
     - --non-interactive
     - --pretty
+    - --python-version=3.8
```

Comment:
This is incorrect, docs say `Type check code assuming it will be running on Python x.y` -- that is not the same as running on that python version and I am sure that the outcome will be different.

If you really want run **with**  different versions of python, you need to tell pre-commit to do so, it is possible to enforce python version.

When you do this also enforce the implicit one to run with minimal python and add a comment like "minimal version of python we support" and add the manual one using the maximal version of python we support (py310), also mentioning it. That will help someone else understand why these values were picked and be able to bump them accordingly.

If you want we could switch the implicit version to run max version instead of min version, I do not have a strong preference for one of them.

Reply:
> This is incorrect, docs say `Type check code assuming it will be running on Python x.y` -- that is not the same as running on that python version and I am sure that the outcome will be different.
> 
> If you really want run **with** different versions of python, you need to tell pre-commit to do so, it is possible to enforce python version.

I'm not sure if this judgment is correct. The approach I used here is derived from other projects I trust and I've seen using it. They've never had to change the runtime. In fact, I've integrated this approach in some projects recently and it was quite evident that it follows through different code paths depending on `--python-version`, requiring different sets of `additional_dependencies`.

Moreover, when you mentioned this, I've gone ahead and checked the documentation of mypy itself too — it requires Python 3.6 in runtime but still supports checking Python 2.7, for example. This supports the point that changing the runtime is not necessary:
> Although you must install Python 3 to run mypy, mypy is fully capable of type checking Python 2 code as well [...]

_(https://mypy.readthedocs.io/en/stable/getting_started.html#installing-and-running-mypy — note at the bottom of the section)_

I suspect this is because MyPy does not evaluate the code in runtime but rather inspects AST statically instead. This also explains how it works with type stubs and type comments.

The bottom line is that it is not required to change `language_version` to make it work correctly.
OTOH, if you have a reproducer that clearly demonstrates what you imply — please share it, and it would be reasonable to reconsider, then. But until then, I've only found evidence that it works as documented.

> When you do this also enforce the implicit one to run with minimal python and add a comment like "minimal version of python we support" and add the manual one using the maximal version of python we support (py310), also mentioning it. That will help someone else understand why these values were picked and be able to bump them accordingly.

This sounds reasonable. Having an explanatory motivation comment is always a good idea in such cases. In fact, there's even a WPS rule (for Python code, of course, but the ideas are similar) that encourages people not to pass arbitrarily chosen constants and have them available under named context-aware variables first.

> If you want we could switch the implicit version to run max version instead of min version, I do not have a strong preference for one of them.

I don't either. Both options have advantages and downsides which I'm not trying to explore in this PR at the moment. But if you have a preference, feel free to document your thoughts here or in a follow-up.

Reply:
(marking as resolved because it's been addressed)